consensus/src/quorum_store/network_listener.rs

```diff
@@ -0,0 +1,123 @@
+    round_manager::VerifiedEvent,
+};
+use aptos_channels::aptos_channel;
+use aptos_logger::debug;
+use aptos_types::PeerId;
+use futures::StreamExt;
+use tokio::sync::mpsc::Sender;
+
+pub(crate) struct NetworkListener {
+    network_msg_rx: aptos_channel::Receiver<PeerId, VerifiedEvent>,
+    batch_reader_tx: Sender<BatchReaderCommand>,
+    proof_coordinator_tx: Sender<ProofCoordinatorCommand>,
+    remote_batch_coordinator_tx: Vec<Sender<BatchCoordinatorCommand>>,
+    proof_manager_tx: Sender<ProofManagerCommand>,
+}
+
+impl NetworkListener {
+    pub(crate) fn new(
+        network_msg_rx: aptos_channel::Receiver<PeerId, VerifiedEvent>,
+        batch_reader_tx: Sender<BatchReaderCommand>,
+        proof_coordinator_tx: Sender<ProofCoordinatorCommand>,
+        remote_batch_coordinator_tx: Vec<Sender<BatchCoordinatorCommand>>,
+        proof_manager_tx: Sender<ProofManagerCommand>,
+    ) -> Self {
+        Self {
+            network_msg_rx,
+            batch_reader_tx,
+            proof_coordinator_tx,
+            remote_batch_coordinator_tx,
+            proof_manager_tx,
+        }
+    }
+
+    pub async fn start(mut self) {
+        debug!("QS: starting networking");
+        //batch fragment -> batch_aggregator, persist it, and prapre signedDigests
+        //Keep in memory caching in side the DB wrapper.
+        //chack id -> self, call PoQSB.
+        while let Some(msg) = self.network_msg_rx.next().await {
+            // debug!("QS: network_listener msg {:?}", msg);
+            match msg {
+                // TODO: does the assumption have to be that network listener is shutdown first?
+                VerifiedEvent::Shutdown(ack_tx) => {
+                    debug!("QS: shutdown network listener received");
+                    ack_tx
+                        .send(())
+                        .expect("Failed to send shutdown ack to QuorumStore");
+                    break;
+                },
+                VerifiedEvent::SignedDigestMsg(signed_digest) => {
+                    // debug!("QS: got SignedDigest from network");
+                    let cmd = ProofCoordinatorCommand::AppendSignature(*signed_digest);
+                    self.proof_coordinator_tx
+                        .send(cmd)
+                        .await
+                        .expect("Could not send signed_digest to proof_coordinator");
+                },
+                VerifiedEvent::FragmentMsg(fragment) => {
+                    counters::DELIVERED_FRAGMENTS_COUNT.inc();
+                    let idx = fragment.source().to_vec()[0] as usize
+                        % self.remote_batch_coordinator_tx.len();
+                    debug!(
+                        "QS: peer_id {:?},  # network_worker {}, hashed to idx {}",
+                        fragment.source(),
+                        self.remote_batch_coordinator_tx.len(),
+                        idx
+                    );
+                    self.remote_batch_coordinator_tx[idx]
```

Comment:
where does the remote_batch_coordinator receive EndBatch?

also if processing remote and local fragments is different, why don't we have two classes - LocalBatchCoordinator and RemoteBatchCoordinator, if they have no overlap?

Reply:
@zekun000 had the same comment. I'll create an item to work on this.